docs/decisions/0010-sfs-versioning.md

```diff
@@ -0,0 +1,114 @@
+
+Object **type** is an enum of Regular, Delete Marker.
+
+## Object State Machine
+
+```mermaid
+stateDiagram-v2
+    [*] --> Open : Create new version
+    Open --> Committed : Success
+    Open --> Open : Writing
+    Committed --> Deleted : Delete
+    Open --> [*] : Failure
+    Deleted --> [*] : Garbage Collection
+```
+
+*Open* - Initial state. Data is in flight. Data on disk is dirty. A
+version may stay in this state if, for example, a client fails during
+a PUT operation. A to be defined GC process cleans this up after a
+time.
+
+*Committed* - Upload finished. Data persisted. GETs and LIST
+will return this object.
+
+*Deleted* - Soft deletion state, but in terms of S3 permanent deletion.
+
+*terminal state* - the row no longer exists in the database.
+
+An object will never move back from Deleted to Committed. Deleting a
+version is treated as permanent, even though it is only permanent after
+the GC made it so.
+
+## Object Version Types
+
+An object version may be of special type *delete marker*, representing
+an S3 delete marker in versioned buckets.
+
+## Garbage Collection
+
+Input: object versions in deleted state.
+
+Either delete single versions or delete whole objects.
+
+In case whole objects, delete the object row if all versions are deleted.
+In a transaction delete all versions, then the object.
+
+Should a concurrent transaction add a new version either will fail on
+the foreign key constraint between version and object. In the GC we
+continue with the next object; in the create path we retry or let the
+client retry.
+
+## Operations
+
+### Deletion
+
+Deleting a version is not to be confused with hiding objects by adding
+deletion markers. Deleting an object sets the state to DELETED.
+
+### Hiding / Deletion Marker
+
+A delete to a versioned bucket without a version creates a delete
+marker. This hides the object and all its version from unversioned GETs
+and LISTS.
+
+SFS implements this with object versions having type delete marker. If
+one exists on an object, it is excluded.
+
+Deletion markers are object versions and follow the same state machine
+and garbage collection as regular versions do. They however don't have
+a size and data on the filesystem.
+
+### Create / Update
+
+(versioned, unversioned buckets) In a transaction:
+
+1. Find or create an object row (id, name, bucket)
+2. Create an object version in state OPEN with creation time of now.
+    If versioned, use the version id provided. Otherwise generate one.
+
+Receive and write data.
+
+On completion:
+(versioned buckets) set state COMMITTED of the previously created version. Record
+checksum, size, etc.
+
+(unversioned buckets) set state COMMITTED of the previously created version
+AND set all other versions DELETED.
+
+### Access: Listing, GETs
+
+Many operations expect a head, null, last or "canonical" version of
+an object. We define this as the *last* version, having the *latest*
+creation time.
```

Comment:
The current implementation uses de `id` of the version. The greater `id` is the latest version. (`id` is the primary key of the versioned object table)
If we want to implement this based on time... Wouldn't it be better to use modification time instead to creation time?. In case that 2 threads are putting a version in parallel, using the creation time means the first trying to write wins, while using the modification time means the *last* that finished writing wins. I think it makes more sense that the one winning is the last writing?

Also, if we want to query what's the latest version based on creation/modification time we might need to change the schema, because right now those time columns are stored as `ceph::real_time` which is stored as a `BLOB` in the database. 

Problem is that _maybe_ the `SQLite` data structures have not enough precision?


**From the SQLite documentation**
```
SQLite does not have a storage class set aside for storing dates and/or times. Instead, the built-in [Date And Time Functions](https://www.sqlite.org/lang_datefunc.html) of SQLite are capable of storing dates and times as TEXT, REAL, or INTEGER values:

TEXT as ISO8601 strings ("YYYY-MM-DD HH:MM:SS.SSS").
REAL as Julian day numbers, the number of days since noon in Greenwich on November 24, 4714 B.C. according to the proleptic Gregorian calendar.
INTEGER as Unix Time, the number of seconds since 1970-01-01 00:00:00 UTC.
```

Reply:
> Also, if we want to query what's the latest version based on creation/modification time we might need to change the schema, because right now those time columns are stored as `ceph::real_time` which is stored as a `BLOB` in the database.

On the multipart work I'm using timestamps to store a `ceph::real_time` instead, which I think is stored as a double. The conversion is trivial to do, and we get to use those fields. I didn't think about the necessary precision though, I should double check that.

Reply:
Looks like we need to standardize on a time format and add that to the high level doc.

For reference: 
`ceph::real_time` is `std::chrono::time_point<real_clock>`
`real_clock` is Ceph's high resolution real time clock. Nanosecond resolution.

I'd love to keep as much resolution as we can. Minimum microseconds. At the same time leveraging sqlite indices, range queries, sorting is important too.

Options: 
1. Store as ISO8601 strings. We loose resolution. Values are monotonically increasing. order by and range queries work nicely. We can use sqlite datetime functions.
2. Store as integer. Unix time is second resolution = meh. So store micos? nanos? SQLite has max 64bit integers. I don't thing struct timespec fits good into this.
3. Store as multiple columns. ISO8601 string + nanoseconds part or two integers. First for search, second for resolution. 
4. Store as blob. I read somewhere[42] that sqlite does comparisons based on memcmp on this. If we encode the time data right, we can have it comparable and monotonically increasing. No idea if we can do BETWEEN queries and how it indexes. Probably the least desirable format. 

What else?


[42] https://www.sqlitetutorial.net/sqlite-data-types/

Reply:
IMO, we should try to keep the same precision that `rgw` has... (nanos). 
I would prefer not to have things in more than 1 column.

Isn't 2 ^ 64 enough? .... it should cover up to year 2554, right?

Reply:
I'm using `ceph::real_clock::to_double(mp.mtime)` to obtain a double, which I assumed would capture a high resolution clock value, but now I'm not so sure. It seems to translate to

`std::chrono::duration<double>(t.time_since_epoch()).count();`

which seems to be using the default for `std::chrono::duration`, that being "seconds" as far as I can tell.

Reply:
then again, we _could_ rely on `std::chrono` directly when converting to/from the database, and apply whatever precision we want.

Reply:
`ceph::real_time` is, if my understanding is correct, represented by a `uint64_t`. 

`ceph::real_time` time_point is defined as: 
```c++
typedef std::chrono::time_point<real_clock> time_point;
```

`ceph::real_clock` is:
```c++
class real_clock {
public:
  typedef timespan duration;
  typedef duration::rep rep;
  typedef duration::period period;
```
and `timespan` is: 
```c++
typedef std::chrono::duration<ceph::rep, std::nano> ceph::timespan
```

and... finally: `ceph:rep`: 
```c++
namespace ceph {
// Currently we use a 64-bit count of nanoseconds.

// We could, if we wished, use a struct holding a uint64_t count
// of seconds and a uint32_t count of nanoseconds.

// At least this way we can change it to something else if we
// want.
typedef uint64_t rep;
```

So the actual `ceph::real_clock` doesn't have the resolution that we can get from `struct timespect` (16 bytes), it fits in 64 bits by design. (I confirmed it printing out sizeof(now), being now = ceph::real_clock::now() )

Reply:
Actually... there is a sqlite_uint64.

https://www.sqlite.org/c3ref/int64.html

and it is also present in `sqliteorm`:
```c++
namespace sqlite_orm {

    using int64 = sqlite_int64;
    using uint64 = sqlite_uint64;
```

So...taking into account that `ceph::real_clock` is represented as a uint64_t,  storing a `ceph::real_clock::time_point` value would be something like:
```c++
std::chrono::duration_cast<std::chrono::nanoseconds>(time_value.time_since_epoch()).count()
```

and reading a `ceph::real_clock::time_point` from the database:
```c++
ceph::real_clock::time_point(std::chrono::nanoseconds(value_from_db));
```

Reply:
meh. I suspected `sqlite` worked internally with signed 64 bits integers and I did a small test.

I stored 5 objects, with the following values in a `test` field:
* 0
* std::numeric_limits<int64_t>::max() - 1 
* std::numeric_limits<uint64_t>::max() - (std::numeric_limits<uint64_t>::max()/4)
* std::numeric_limits<uint64_t>::max()
* 1234555555

I did the following query:
```c++
auto objs = storage.get_all<DBObjs>(
        sqlite_orm::order_by(&DBObjs::test).desc()
    );
```
and the result is:
```
9223372036854775806
1234555555
0
18446744073709551615
13835058055282163712
```

so sign is taken into account.
I declared the `test` field as `sqlite_orm::uint64`

I had no issue storing and retrieving uint64 values... but the problem is when we try to get things in order... or mix/max values.

I also tried to do 
```c++
storage.max(&DBObjs::test);
```
and it returns the maximum signed 64 bits integer: 9223372036854775806

I'll investigate a bit more... with signed 64 bits numbers we can maybe work up to 63 bits which would give us time values from epoch until year 2262 (if I'm not mistaken), maybe the `s3gw` is still in use by that year... who knows... :)

Reply:
I've been testing another possible solution. I prefer the uint64 storage, but as observed above we can't rely on it. (and I suspect we can loose precision storing a `double`)

So... for storing the timestamp... I tested to store the hex string representation of the uint64_t that represents the `ceph::real_clock::time_point` (max  uint64 would be stored as `ffffffffffffffff`, 0 would be represented as `0000000000000000`.
That means... fixed length strings  that `sqlite` is able to sort and compare and so on.

Obviously, in terms of latency, this option is slower because there is a layer of conversion, but strings will be always 16 char strings. 

For testing it I used this for the uint64_t -> string conversion:
```c++
auto db_value = fmt::format("{:0>16x}", uint64_value);
```

I saw fmt::scan is still in the test folder of libfmt so I simply used `std::stoull` for the str to uint64_t conversion.
```c++
auto rgw_value = std::stoull(str_db_value, nullptr, 16);
```

This time the test was correct calling the same `order_by` query:
```
18446744073709551615
13835058055282163712
9223372036854775806
1234555555
0
```

and `storage.max()` returns 18446744073709551615, which is correct.

Reply:
> And finally... something I wanted to investigate long ago when all this started and it's custom type binding in `sqlite_orm`

Very cool! Wanted to try the binding support for a while as well :). Looks good. It could replace the conversion layer we have if we add bindings for the ceph types, right?

Reply:
This would move the conversion to another place, but the conversion code will be still needed.
We won't see any performance gain. (and the code to be maintained will be highly templated code, which not everybody loves)

This is something I wanted to investigate since the very beginning but given that we didn't had the time I could not go deeper. 

I think it makes user code cleaner, because we work with our own types.

Anyway, I would maybe do it for the timestamps in this iteration and will try to change the rest when the versioning stuff is done. I don't think this is as urgent as versioning work.

Reply:
As I know there is a lot of noise I'm making a summary:

Options I see so far.
* Store timestamps as `int64` (which will limit our timestamps range up to year 2262)
* Store timestamps as 16 chars hex string (we can store the whole uint64 range and, if in the future upstream adds more precisions we can increase the number of chars... This comes with a conversion code)
* Store as `uint64` (but we cannot query by timestamps because sqlite takes sign into account for order_by and max, main)


And... for deciding how to identify what's the last version of an object.
* Max creation time
* Max modification time (or another field that we create so the last thread writing wins)
* Max version id (I think we can go for this in favor of max creation time, because they are equivalents and we don't need to change the timestamps code. When we set creation time we also create a new version id, so the greater one will still be the latest one)

Reply:
I'd like to add option int64, but with microsecond resolution (max 2554-07-21 23:34:33.709551).